### ISSUE TITLE
Export Network to SVG picture

### ISSUE THREAD

**Issue by TRechkalov** [almende/vis#1784]
Hello!

I want to export Network to SVG picture. By default it is not possible.
I tried several ways but no one helps:
1. Canvas2Svg - I replaced canvas 2d context by Canvas2Svg and added extra methods for drawing circle but it draw only edges.
2. Export to Graphviz with fixed positions also doesn't work for me.

Maybe you have some new ideas instead making Graphviz and d3.js export manually?

**Comment by dragonzone** [comment]
I have used vis.js Network w/ GraphViz and works perfectly. It took me a while to figure it out, but you need to use "dot" w/ these params. Here is the full command I used:

dot -n2 -Kneato -Tpdf ./test.dot -o ./test.pdf

**Comment by dragonzone** [comment]
Well, for my case, it was to convert to PDF, but you can easily change it to SVG w/o any problem. Though I have not tried that, but should be similar path.

**Comment by ragol** [comment]
@dragonzone how do you produce the dot file from vis.js?

**Comment by dragonzone** [comment]
Here is the pseudo code for how I did it:

1) Vis.js Network is drawn using the DataSet module, you can go here to read more on it (http://visjs.org/docs/data/dataset.html)
2) You take all the network's nodes & edges, convert them to JSON
3) Make AJAX call to server to POST the JSON's network's nodes & edges data
4) On server side, I used Java, but you can use any server-side language to convert that object to a DOT format. You can go here to see some DOT examples (http://graphs.grevian.org/example)
5) On server, once the DOT file is generated, you call the command in my earlier discussion, which will generate a PDF for any other supported file format from GraphViz on to the server temp folder
6) You take the generated PDF file and send it back to the user.

Hope this helps!

**Comment by mvivekc** [comment]
Conversion of current graph displayed to SVG/PNG would be a great feature. Was evaluating Cytoscape/visjs and the main feature where cytoscape stands out is with the export feature. Image download is provided by highcharts and other major chart libs. Adding this would be big boost in visjs's feature set,

**Comment by dragonzone** [comment]
Vis.js uses HTML5 canvas to draw the network. If you just need to create PNG image, you can just use HTML5 Canvas drawImage() (http://www.w3schools.com/html/tryit.asp?filename=tryhtml5_canvas_tut_img), to draw the network. I have done that awhile back than it worked. However, it is only in PNG format. With DOT, it has much more flexibility to the format you want to export to 20+ formats. Go here for complete list: http://soc.if.usp.br/manual/graphviz/html/info/output.html.

**Comment by mvivekc** [comment]
I agree, it should be simple. With DOT - u need some external server running. Since the export as PNG is simple, visjs itself can accommodate for it. network.exportAs("png")  |  network.exportAs("svg")  should not be too complicated to achieve out of the box.

**Comment by ragol** [comment]
@dragonzone So basically you wrote a visjs-to-dot converter yourself, right? I thought, I have overlooked something and vis.js could do that on its own.

**Comment by dragonzone** [comment]
@ragol , yes, I had to convert it myself, which is very easy to do.

**Comment by anz000** [comment]
Hi @dragonzone , could you give more info on 1. converting nodes/edges to JSON and 2. convert to DOT? Will the visualization look be preserved?

**Comment by dragonzone** [comment]
You take the vis.js network nodes & edges and convert them to DOT format. An example of DOT file format is this: http://graphs.grevian.org/example#example-4

When you export, you need to set the X & Y coordinate in vis.js and add them to your DOT file so it will preserve the exact look that is on vis.js network.

**Comment by anz000** [comment]
Great Info. I played around with DOT format, and converted to vector pdf, I believe it meets my requirement. 

I'm currently using visNetwork in R, so I don't know if I could get all X,Y co-ordinate to recreate exactly like vis.js graph. How can I get the XY co-ordinate and integrate that information into the DOT file?

Thanks in advance.

**Comment by dragonzone** [comment]
I have not tried, but I think once you create the vis.js network, you will have a reference to the network object. From there, you can use that network reference to get all the nodes and edges and their coordinates. I have written a quick sample on how to get the nodes X & Y coordinates from there onward for another issue awhile back: https://github.com/almende/vis/issues/2017

**Comment by anz000** [comment]
Thank you for redirecting me to the right direction. 

The graph looks good. I was playing around with managing edges as in Vis. The "splines=curved" settings curves the edges too much, but I guess I can live with that.

**Comment by wimrijnders** [comment]
Hi there! This issue has been labeled `Inactive`, because there has been no activity here for over two months. If the inactivity continues, the issue will be closed in two weeks.

If you feel that the issue should remain open, please comment here, preferably with motivation. Thanks!

**Comment by wimrijnders** [comment]
Closed due to inactivity. Feel free to reopen if this is still relevant.

**Comment by dockstreet** [comment]
I would like to re-open this issue to investigate downstream usage for this:

**Comment by wimrijnders** [comment]
Be my guest.

**Comment by NtsDK** [comment]
A long time ago I tried to replace network canvas with fabricjs. It was without success but it is one of the possible ways to solve this problem.

**Comment by sancarn** [comment]
+1 This would be an extremely useful feature!

**Comment by NtsDK** [comment]
One more possible workaround - extracting and saving graph in graphML format for yEd. yEd editor can be used to save svg.

**Comment by SumNeuron** [comment]
Exporting to DOT is doable, but a nuisance if you want to copy all the custom styles and groups you might have applied (not to mentioned inheritance). Since the graph2d module has support for SVG I hope this issue could get an official resolution soon.
